### PR TITLE
Add function-no-unknown

### LINKED ISSUE #5853: Add function-no-unknown
### What is the problem you're trying to solve?

I'm looking for a list of known CSS functions in order to catch typo errors.

see https://github.com/stylelint-scss/stylelint-scss/issues/581#issuecomment-1017484418

### What solution would you like to see?

1. Stylelint could export a list of known CSS functions that could be imported inside stylelint.config.js
2. Stylelint could provide a rule `function-no-unknown` which throws errors for unknown functions

### PR DESCRIPTION
**Pull request by gnuletik**
> Which issue, if any, is this issue related to?

Closes #5853

> Is there anything in the PR that needs further explanation?

"No, it's self-explanatory."

### CODE REVIEW

**Comment by gnuletik** [comment]
There's an issue on Windows builds when loading [`css-functions-list`](https://www.npmjs.com/package/css-functions-list):
>     ENOENT: no such file or directory, open '/D:/a/stylelint/stylelint/node_modules/css-functions-list/cjs/index.json'

Does someone use Windows and have an idea to fix it?

**Comment by ybiquitous** [comment]
@gnuletik How about using `URL` as below?

```js
fs.readFileSync(new URL("file://" + listPath))
```

File: lib/rules/function-no-unknown/index.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
+'use strict';
+
+const fs = require('fs');
+const valueParser = require('postcss-value-parser');
+const functionsListPath = /** @type {string} */ (
+	/** @type {unknown} */ (require('css-functions-list'))
```
Comment (alexander-akait) [inline lib/rules/function-no-unknown/index.js:6]: This package is too old and doesn't contain all functions

Comment (ybiquitous) [reply lib/rules/function-no-unknown/index.js:6]: Oh, do you know some alternative or another way?

Comment (ybiquitous) [reply lib/rules/function-no-unknown/index.js:6]: If the package is insufficient, I think we should contain a function name set internally like the math functions: 🤔 

https://github.com/stylelint/stylelint/blob/b0523d8e76929d8c7224f884b0b256c2657ce977/lib/reference/mathFunctions.js#L3

Any thoughts?

Comment (gnuletik) [reply lib/rules/function-no-unknown/index.js:6]: This package uses data from MDN.
I cloned it locally and run `npm run generate-list` and got 10 new functions in the diff.
I guess we can make a PR to update the package, unless you want to maintain it inside stylelint. Up to you :)

Comment (ybiquitous) [reply lib/rules/function-no-unknown/index.js:6]: @gnuletik Thanks for the feedback. Could you please open a PR to [niksy/css-functions-list](https://github.com/niksy/css-functions-list), please?
If the repository owner will maintain the package, we can go ahead.

If not, then we will maintain the function list by ourselves. I'm a bit concerned about our maintenance resources... 🤔

Comment (gnuletik) [reply lib/rules/function-no-unknown/index.js]: > @gnuletik How about using `URL` as below?
> 
> ```js
> fs.readFileSync(new URL("file://" + listPath))
> ```

I just made the change. I need approval to run the CI.

**Comment by gnuletik** [comment]
> @gnuletik How about using `URL` as below?
> 
> ```js
> fs.readFileSync(new URL("file://" + listPath))
> ```

I just made the change. I need approval to run the CI.

File: lib/rules/function-no-unknown/index.js (hunk 1 above)
Comment (alexander-akait) [reply lib/rules/function-no-unknown/index.js]: I think we can have list here, it is not hard to maintenance

Comment (gnuletik) [reply lib/rules/function-no-unknown/index.js]: Done :) https://github.com/niksy/css-functions-list/pull/1

File: docs/user-guide/rules/list.md [hunk 2]
```diff
@@ -120,6 +120,7 @@ Within each cateogory, the rules are grouped by the [_thing_](http://apps.workfl
 
 - [`function-allowed-list`](../../../lib/rules/function-allowed-list/README.md): Specify a list of allowed functions.
 - [`function-disallowed-list`](../../../lib/rules/function-disallowed-list/README.md): Specify a list of disallowed functions.
+- [`function-no-unknown`](../../../lib/rules/function-no-unknown/README.md): Disallow unknown functions.
```
Comment (ybiquitous) [inline docs/user-guide/rules/list.md:123]: Should the rule be in the "Avoid errors" section?

File: lib/rules/function-no-unknown/index.js (hunk 1 above)
Comment (niksy) [reply lib/rules/function-no-unknown/index.js]: New version has been pushed! 😉

Comment (ybiquitous) [reply lib/rules/function-no-unknown/index.js]: @niksy Thank you so much! 😊

File: lib/rules/function-no-unknown/index.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+		root.walkDecls((decl) => {
+			const { value } = decl;
+
+			valueParser(value).walk((node) => {
+				if (node.type !== 'function') {
+					return;
+				}
+
+				// custom function: https://github.com/w3c/css-houdini-drafts/issues/1007
+				if (node.value.startsWith('--')) {
```
[the remark this answers is not in the document]
Comment (gnuletik) [reply lib/rules/function-no-unknown/index.js]: Yep! Done.

File: docs/user-guide/rules/list.md (hunk 2 above)
Comment (gnuletik) [reply docs/user-guide/rules/list.md]: That makes sense! Done.

File: lib/utils/__tests__/isCustomFunction.test.js [hunk 4]
```diff
@@ -0,0 +1,12 @@
+'use strict';
+
+const isCustomFunction = require('../isCustomFunction');
+
+it('isCustomMediaQuery', () => {
```
Comment (ybiquitous) [inline lib/utils/__tests__/isCustomFunction.test.js:5]: [typo]

```suggestion
it('isCustomFunction', () => {
```

File: package.json [hunk 5]
```diff
@@ -111,6 +111,7 @@
     "balanced-match": "^2.0.0",
     "colord": "^2.9.2",
     "cosmiconfig": "^7.0.1",
+    "css-functions-list": "^2.0.0",
```
Comment (ybiquitous) [inline package.json:114]: [suggest] Bump up, please?

```suggestion
    "css-functions-list": "^3.0.0",
```

https://github.com/niksy/css-functions-list/releases/tag/v3.0.0

**Review by ybiquitous — COMMENTED** [review]
@gnuletik Thanks, almost good! 👍🏼

**Review by ybiquitous — APPROVED** [review]
Thanks, LGTM 👍🏼

**Review by jeddy3 — CHANGES_REQUESTED** [review]
Thanks for the pull request.

I've requested some changes. Let's also add an accept test for `TRANSFORM` and an reject one for `UNKNOWN`.

**Comment by gnuletik** [comment]
Thanks @jeddy3 !
I applied your changes, added some test for the `if (!isStandardSyntaxFunction) { ... }` branch and rebased.

File: lib/rules/function-no-unknown/__tests__/index.js [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+	ruleName,
+	config: false,
+	skipBasicChecks: true,
+
+	accept: [
+		{
+			code: 'a { transform: shouldskip(4); }',
+		},
+	],
+});
```
[the remark this answers is not in the document]
Comment (gnuletik) [reply lib/rules/function-no-unknown/__tests__/index.js]: However, this removal generates a test coverage warning : https://github.com/stylelint/stylelint/pull/5865/checks?check_run_id=5069375681

[the remark this answers is not in the document]
Comment (jeddy3) [reply lib/rules/function-no-unknown/__tests__/index.js]: I think it's ok to have that warning.

File: lib/rules/function-no-unknown/README.md [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+<!-- prettier-ignore -->
+```css
+a { transform: unknown(1); }
+/**            ↑
+ * Functions like this */
+```
+
+This rule considers functions defined in the CSS Specifications to be known.
+
+This rule ignores double-dashed custom functions, e.g. --custom-function().
```
Comment (jeddy3) [inline lib/rules/function-no-unknown/README.md:14]:
```suggestion
This rule ignores double-dashed custom functions, e.g. `--custom-function()`.
```

Minor nit.

File: lib/rules/function-no-unknown/__tests__/index.js [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+		},
+		{
+			code: 'a { transform: --custom-function(1px); }',
+		},
+		{
+			code: 'a { transform: scale(0.5) translate(-100%, -100%); }',
+		},
+		{
+			code: 'a { $list: (list) }',
+		},
```
Comment (jeddy3) [inline lib/rules/function-no-unknown/__tests__/index.js:22]: Let's move this SCSS into its own `testRule` that uses the `postcss-scss` syntax as we strive to keep standard and non-standard syntax separate.

**Review by jeddy3 — CHANGES_REQUESTED** [review]
Thanks for making the changes.

I've requested a couple more.

Let's also add an:
- accept test for `TRANSFORM` 
- an reject one for `UNKNOWN`

So that we test that the rule is case insensitive.

File: lib/rules/function-no-unknown/__tests__/index.js [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+			line: 1,
+			column: 23,
+		},
+	],
+});
+
+testRule({
+	ruleName,
+	config: true,
+	skipBasicChecks: true,
```
Comment (jeddy3) [inline lib/rules/function-no-unknown/__tests__/index.js:50]:
```suggestion
	ruleName,
	config: true,
	customSyntax: "postcss-scss",
	skipBasicChecks: true,
```

Let's specify the SCSS syntax (even though the default parser can parse the construct).

**Review by jeddy3 — CHANGES_REQUESTED** [review]
Thanks for making the changes.

One last suggestion and I think we're ready to merge.

**Review by jeddy3 — APPROVED** [review]
Thanks for making the changes. LGTM.

**Comment by jeddy3** [comment]
Changelog entry:

> - Added: `function-no-unknown` rule ([#5865](https://github.com/stylelint/stylelint/pull/5865)).

**Comment by torhovland** [comment]
This is now starting to complain when I have

```
@use 'sass:math';

.service-detail-header-logo__active-icon {
  top: math.div(-$pulse-size, 4);
  ...
```

```
Unexpected unknown function "math.div"  function-no-unknown
```

**Comment by ybiquitous** [comment]
@torhovland It is an issue of `stylelint-scss`: ⬇️ 

https://github.com/stylelint-scss/stylelint-scss/issues/581

**Comment by ybiquitous** [comment]
As a workaround, you can [disable the rule](https://stylelint.io/user-guide/configure/#overrides) like this:

```json
{
  "overrides": [
    {
      "files": ["*.scss", "**/*.scss"],
      "rules": {
        "function-no-unknown": null
      }
    }
  ]
}
```

**Comment by kristerkari** [comment]
The SCSS version is now also available: 
https://github.com/stylelint-scss/stylelint-scss/releases/tag/v4.2.0

**Pull request merged**